File: apps/gals_heat/create_gals_heat_instance.py
#!/usr/bin/env python3

from graph.core import *

from graph.load_xml import load_graph_types_and_instances
from graph.save_xml_stream import save_graph

import sys
import os
import math
import random

import os
appBase=os.path.dirname(os.path.realpath(__file__))

src=appBase+"/gals_heat_graph_type.xml"
(graphTypes,graphInstances)=load_graph_types_and_instances(src,src)

urand=random.random

n=16
if len(sys.argv)>1:
    n=int(sys.argv[1])

assert n>=3

h=1.0/n
alpha=1

dt=h*h / (4*alpha) * 0.5


assert h*h/(4*alpha) >= dt

leakage=1

weightOther = dt*alpha/(h*h)
weightSelf = (1.0 - 4*weightOther)

weightOther = weightOther * leakage

graphType=graphTypes["gals_heat"]
devType=graphType.device_types["cell"]
dirichletType=graphType.device_types["dirichlet_variable"]
exitNodeType=graphType.device_types["exit_node"]

instName="heat_{}_{}".format(n,n)

properties={}

res=GraphInstance(instName, graphType, properties)

nodes={}

for x in range(0,n):
    sys.stderr.write(" Devices : Row {} of {}\n".format(x, n))
    for y in range(0,n):
        meta={"loc":[x,y],"hull":[ [x-0.5,y-0.5], [x+0.5,y-0.5], [x+0.5,y+0.5], [x-0.5,y+0.5] ]}
        edgeX = x==0 or x==n-1
        edgeY = y==0 or y==n-1
        if x==n//2 and y==n//2:
            props={ "bias":0, "amplitude":1.0, "phase":1.5, "frequency": 100*dt, "neighbours":4 }
            di=DeviceInstance(res,"v_{}_{}".format(x,y), dirichletType, props, None, meta)
            nodes[(x,y)]=di
            res.add_device_instance(di)
        elif edgeX != edgeY:
            props={ "bias":0, "amplitude":1.0, "phase":1, "frequency": 70*dt*((x/float(n))+(y/float(n))), "neighbours":1 }
            di=DeviceInstance(res,"v_{}_{}".format(x,y), dirichletType, props, None, meta)
            nodes[(x,y)]=di
            res.add_device_instance(di)
        elif not (edgeX or edgeY):
            props={ "iv":urand()*2-1, "nhood":4, "wSelf":weightSelf }
            di=DeviceInstance(res,"c_{}_{}".format(x,y), devType, props, None, meta)
            nodes[(x,y)]=di
            res.add_device_instance(di)
            
def add_channel(x,y,dx,dy):
    dst=nodes[ (x,y) ]
    src=nodes[ ( (x+dx+n)%n, (y+dy+n)%n ) ]
    if dst.device_type.id=="cell":
        props={"w":weightOther}
    else:
        props=None  
    ei=EdgeInstance(res,dst,"in", src,"out", props)
    res.add_edge_instance(ei)

for x in range(0,n):
    sys.stderr.write(" Edges : Row {} of {}\n".format( x, n))
    for y in range(0,n):
        edgeX = x==0 or x==n-1
        edgeY = y==0 or y==n-1
        if edgeX and edgeY:
            continue

        if y!=0 and not edgeX:
            add_channel(x,y, 0, -1)
        if x!=n-1 and not edgeY:
            add_channel(x,y, +1, 0)
        if y!=n-1 and not edgeX:
            add_channel(x,y, 0, +1)
        if x!=0 and not edgeY:
            add_channel(x,y, -1, 0)        

finished=DeviceInstance(res, "finished",exitNodeType, {"fanin":len(nodes)}, None) 
res.add_device_instance(finished)

for (id,di) in nodes.items():
    res.add_edge_instance(EdgeInstance(res,finished,"done",di,"finished"))

save_graph(res,sys.stdout)        

File: apps/gals_heat/create_gals_heat_instance.cpp
#include "graph.hpp"
#include "graph_persist_sax_writer.hpp"

#include "rapidjson/document.h"

#include "gals_heat.graph.hpp"

#include <string>
#include <cstdlib>

int main(int argc, char *argv[])
{
  RegistryImpl registry;

  if(argc<2){
    fprintf(stderr, "usage : create gals_heat_instance dstFile [n]\n");
    fprintf(stderr, "  Note that dstFile can't be - at the moment.\n");
    exit(1);
  }

    auto writer=createSAXWriterOnFile(argv[1]);
  
  GraphTypePtr graphType=registry.lookupGraphType("gals_heat");

  unsigned exportDeltaMask=1023;

  unsigned n=16;
  if(argc>2){
    n=strtol(argv[2],0,0);
  }

  assert(n>=3);

  double h=1.0/n;
  double alpha=1;
    
  double dt=h*h / (4*alpha) * 0.5;
  assert(h*h/(4*alpha) >= dt);

  double weightOther = dt*alpha/(h*h);
  double weightSelf = (1.0 - 4*weightOther);

  auto cellType=graphType->getDeviceType("cell");
  auto dirichletType=graphType->getDeviceType("dirichlet_variable");
  
  std::string instName="heat_"+std::to_string(n)+"_"+std::to_string(n);

  DataPtr<gals_heat_properties_t> graphProperties(new gals_heat_properties_t());
  graphProperties->maxTime=n;
  graphProperties->exportDeltaMask=exportDeltaMask;

  auto xyToIndex=[=](unsigned x, unsigned y)
    { return x*n+y; };
  std::vector<std::pair<uint64_t,bool> > nodes(n*n);// Used to hold mapping to writers view of node ids, plus true=dirichlet, false=cell

  writer->onGraphType(graphType);

  auto gInst=writer->onBeginGraphInstance(graphType, instName, graphProperties, rapidjson::Document());

  writer->onBeginDeviceInstances(gInst);

  for(unsigned x=0;x<n;x++){
    fprintf(stderr," Devices : Row %d of %d\n", x, n);
    for(unsigned y=0;y<n;y++){
      double nativeLocation[2]={(double)x,(double)y};
      bool edgeX = (x==0) || (x==n-1);
      bool edgeY = (y==0) || (y==n-1);

      if(edgeX && edgeY)
	      continue;

      rapidjson::Document meta;
      rapidjson::Value loc(rapidjson::kArrayType);
      loc.PushBack(x, meta.GetAllocator());
      loc.PushBack(y, meta.GetAllocator());
      meta.AddMember("loc", loc, meta.GetAllocator());
      
      if(x==n/2 && y==n/2){
        auto props=make_data_ptr<dirichlet_variable_properties_t>();
        props->bias=0;
        props->amplitude=1;
        props->phase=1.5;
        props->frequency=100*dt;
        props->neighbours=4;
        std::string id="v_"+std::to_string(x)+"_"+std::to_string(y);
        uint64_t index=writer->onDeviceInstance(gInst, dirichletType, id, props, std::move(meta));
        nodes[xyToIndex(x,y)]=std::make_pair(index,true);
        //fprintf(stderr, "  (%d,%d) -> %llu\n", x,y, index);
      }else if(edgeX != edgeY){
        auto props=make_data_ptr<dirichlet_variable_properties_t>();
        props->bias=0;
        props->amplitude=1;
        props->phase=1;
        props->frequency=70*dt*(x/(double)n + y/(double)n);
        props->neighbours=1;
        std::string id="v_"+std::to_string(x)+"_"+std::to_string(y);
        uint64_t index=writer->onDeviceInstance(gInst, dirichletType, id, props, std::move(meta));
        nodes[xyToIndex(x,y)]=std::make_pair(index,true);
        //fprintf(stderr, "  (%d,%d) -> %llu\n", x,y, index);
      }else{
        auto props=make_data_ptr<cell_properties_t>();
        props->iv=drand48()*2-1;
        std::string id="c_"+std::to_string(x)+"_"+std::to_string(y);
        uint64_t index=writer->onDeviceInstance(gInst, cellType, id, props, std::move(meta));
        nodes[xyToIndex(x,y)]=std::make_pair(index,true);
        //fprintf(stderr, "  (%d,%d) -> %llu\n", x,y, index);
      }
    }
  }

  writer->onEndDeviceInstances(gInst);

  auto dirichletIn=dirichletType->getInput("in");
  auto cellIn=cellType->getInput("in");
  auto dirichletOut=dirichletType->getOutput("out");
  auto cellOut=cellType->getOutput("out");
  
  auto add_channel=[&](unsigned x,unsigned y ,unsigned dx,unsigned dy)
    {
      auto dst=nodes[ xyToIndex(x,y) ];
      auto src=nodes[ xyToIndex( (x+dx+n)%n, (y+dy+n)%n ) ];

      //      fprintf(stderr,"(%d,%d):%u <- (%d,%d):%u\n", x,y, dst.first, (x+dx+n)%n, (y+dy+n)%n, src.first);

      writer->onEdgeInstance
      (
        gInst,
        dst.first, dst.second ? dirichletType : cellType, dst.second ? dirichletIn : cellIn,
        src.first, src.second ? dirichletType : cellType, src.second ? dirichletOut : cellOut,
        TypedDataPtr()
       );
    };
  
  writer->onBeginEdgeInstances(gInst);
  
  for(unsigned x=0; x<n; x++){
    fprintf(stderr, " Edges : Row %d of %d\n", x,n);
    for(unsigned y=0; y<n; y++){
      bool edgeX = (x==0) || (x==n-1);
      bool edgeY = (y==0) || (y==n-1);
      if(edgeX && edgeY)
	continue;

      if(y!=0 && !edgeX)
	add_channel(x,y, 0, -1);
      if(x!=n-1 && ! edgeY)
	add_channel(x,y, +1, 0);
      if(y!=n-1 && ! edgeX)
	add_channel(x,y, 0, +1);
      if(x!=0 && ! edgeY)
	add_channel(x,y, -1, 0);
    }
  }

  writer->onEndEdgeInstances(gInst);

  writer->onEndGraphInstance(gInst);

  return 0;
}
